src/frontend/mame/ui/barcode.cpp

```diff
@@ -131,26 +131,14 @@ void menu_barcode_reader::handle()
 		case IPT_SPECIAL:
 			if (get_selection_ref() == ITEMREF_NEW_BARCODE)
 			{
-				auto const buflen = std::strlen(m_barcode_buffer);
-
-				// if it's a backspace and we can handle it, do so
-				if ((event->unichar == 8) || (event->unichar == 0x7f))
-				{
-					if (0 < buflen)
-						*const_cast<char *>(utf8_previous_char(&m_barcode_buffer[buflen])) = 0;
-				}
-				else if ((event->unichar >= '0') && (event->unichar <= '9'))
-				{
-					event->append_char(m_barcode_buffer, buflen);
-				}
-				reset(reset_options::REMEMBER_POSITION);
+				if (input_character(m_barcode_buffer, event->unichar, uchar_is_digit))
+					reset(reset_options::REMEMBER_POSITION);
```

Comment:
`event->append_char` limited the length of the input string to the array length of the buffer.  Barcodes were limited to 19 characters plus NUL by this.  Are you sure nothing depends on the length being limited?